theano/gof/wrapper.py

```diff
@@ -0,0 +1,443 @@
+    npy_float64    number = param->attr2;
+    /* You won't need to free them or whatever else. */
+
+
+See :class:`QuadraticOpFunc` and :class:`QuadraticCOpFunc` in ``theano/gof/tests/test_wrapper.py``
+for complete working examples.
+
+"""
+
+from __future__ import absolute_import, print_function, division
+import re
+import hashlib
+from theano.gof.utils import MethodNotDefined, c_cpp_keywords
+from theano.gof import Type
+
+
+class Wrap(dict):
+    """
+    Internal convenient class to wrap many Python objects into one
+    (this class is not safe as the hash method does not check if values are effectively hashable).
+
+    **Example:**
+
+    .. code-block:: python
+
+        from theano.gof import Wrapper, Wrap
+        from theano.scalar import Scalar
+        # You must create a Wrapper first:
+        wp = Wrapper(attr1=Scalar('int32'), key2=Scalar('float32'), field3=Scalar('int64'))
+        # Then you can create a Wrap with the wrapper defined above and values for attributes.
+        w = Wrap(wp, attr1=1, key2=2.0, field3=3)
+        print(w.attr1, w.key2, w.field3)
+        d = dict(attr1=1, key2=2, field3=-1)
+        w2 = Wrap(wp, **d)
+        print(w2.attr1, w2.key2, w2.field3)
+
+    """
+
+    def __init__(self, wrapper, **kwargs):
+        if not isinstance(wrapper, Wrapper):
+            raise TypeError('Wrap: 1st constructor argument should be a Wrapper.')
+        for field in wrapper.fields:
+            if field not in kwargs:
+                raise TypeError('Wrap: Wrapper attribute "%s" not in Wrap args.' % field)
+        super(Wrap, self).__init__(**kwargs)
+        self.__dict__.update(wrapper=wrapper)
+
+    def __repr__(self):
+        return 'Wrap(%s)' % ', '.join([('%s:%s' % (k, type(self[k]))) for k in sorted(self.keys())])
+
+    def __getattr__(self, key):
+        if key not in self:
+            raise AttributeError('Wrap: attribute "%s" does not exist.' % key)
+        return self[key]
+
+    def __setattr__(self, key, value):
+        raise NotImplementedError('Wrap is immutable')
+
+    def __setitem__(self, key, value):
+        raise NotImplementedError('Wrap is immutable')
+
+    def __delitem__(self, key):
+        raise NotImplementedError('Wrap is immutable')
+
+    def __hash__(self):
+        return hash((type(self), self.wrapper) + tuple(
+            # NB: Wrapped data should have been already filtered.
+            self.wrapper.types[i].make_constant(self[self.wrapper.fields[i]]).signature()
+            for i in range(self.wrapper.length)
+        ))
+
+    def __eq__(self, other):
+        return (type(self) == type(other) and self.wrapper == other.wrapper and all(
+            # NB: Wrapped data should have been already filtered.
+            self.wrapper.types[i].values_eq(self[self.wrapper.fields[i]], other[self.wrapper.fields[i]])
+            for i in range(self.wrapper.length)
+        ))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class Wrapper(Type):
```

Comment:
About this, I was thinking to `ParamsType` (with `s` at end of `Params`) as a new name, as it should become the main class to use to provide multiple parameters for an Op !

What do you prefer ?